test-data/unit/check-classes.test

```diff
@@ -4223,3 +4224,35 @@ class C(Any):
         reveal_type(self.bar().__name__) # E: Revealed type is 'builtins.str'
 [builtins fixtures/type.pyi]
 [out]
+
+[case testClassDecoratorIsTypeChecked]
+from typing import Callable, Type
+def decorate(x: int) -> Callable[[type], type]:  # N: "decorate" defined here
+    ...
+def decorate_forward_ref() -> Callable[[Type[A]], Type[A]]:
+    ...
+@decorate(y=17)    # E: Unexpected keyword argument "y" for "decorate"
+@decorate()        # E: Too few arguments for "decorate"
+@decorate(22, 25)  # E: Too many arguments for "decorate"
+@decorate_forward_ref()
+@decorate(11)
+class A: pass
+
+@decorate  # E: Argument 1 to "decorate" has incompatible type "Type[A2]"; expected "int"
+class A2: pass
+
+[case testClassDecoratorIncorrect]
+def not_a_class_decorator(x: int) -> int: ...
+@not_a_class_decorator(7)  # E: "int" not callable
+class A3: pass
+
+not_a_function = 17
+@not_a_function()  # E: "int" not callable
```

Comment:
Also add a test for directly applying `not_a_function`:
```python
@not_a_function
class C: ...
```

Reply:
Done